Calypso Build Package: Add cssChunkFilename arg to SASS shard, compute from JS counterpart

ockham opened the pull request:
#### Changes proposed in this Pull Request

Follow-up to @sirreal's #32192, based on conversation in https://github.com/Automattic/jetpack/pull/12014.

- Add cssChunkFilename arg to SASS shard
- Compute cssFilename and cssChunkFilename from their JS counterparts

The rationale here is that we want our CSS filenames and chunk filenames to reflect their JS counterparts. Fortunately, `mini-css-extract-plugin` supports [options for both](https://github.com/webpack-contrib/mini-css-extract-plugin#minimal-example).

We thus add a `cssChunkFilename` arg to the SASS shard. Furthermore, we change `calypso-build`'s webpack config to compute the CSS filenames (and chunk filenames) from webpack's [`output-filename` and `output-chunk-filename`](https://webpack.js.org/api/cli/#output-options) args.

I'm not a big fan of "being smart" (in this case: determining the CSS filenames from the JS counterparts by removing the `.js`, and adding the `.css` suffix), but it seems like the best option in this case, since
1. It's a reasonable assumption that filenames for both cases should follow the same pattern
2. It's a reasonable assumption that built JS files are suffixed `.js`
3. While webpack supports `output-filename` and `output-chunk-filename` args natively, the CSS counterparts are obviously specific to `mini-css-extract-plugin`.
4. However, they're obviously similar in spirit -- and I don't think it makes sense to add those non-native options as args to `getWepackConfig()`'s signature.

#### Testing instructions

Build `o2-blocks` - no changes. `npx lerna run build`

#### Sidenote

This is also going to come in handy for #32083.

ockham committed:
Calypso Build: Add cssChunkFilename arg to SASS shard

ockham committed:
Calypso Build: Compute cssFilename and cssChunkFilename from their JS counterparts

ockham committed:
Calypso Build: Bump package version

sirreal commented on packages/calypso-build/webpack.config.js:
```diff
@@ -107,7 +110,7 @@ function getWebpackConfig(
 			} ),
 			new webpack.IgnorePlugin( /^\.\/locale$/, /moment$/ ),
-			...SassConfig.plugins( { cssFilename, minify: ! isDevelopment } ),
+			...SassConfig.plugins( { cssChunkFilename, cssFilename, minify: ! isDevelopment } ),
```
This `css*` pattern seems odd to me.

```suggestion
			...SassConfig.plugins( {
				chunkFilename: cssChunkFilename,
				filename: cssFilename,
				minify: ! isDevelopment,
			} ),
```

simison replied:
I agree, it's clear to have `cssFilename` outside the shard but inside the shard it's the only "filename" so makes sense to drop `css*`.

ockham replied:
I agree, too. I didn't want to change the shard's interface with this PR since I think some consumers use it directly, so it constitutes a breaking API change.

sirreal replied:
It's beta! 🤷‍♂️ 😅 😬

sirreal commented on packages/calypso-build/webpack/sass.js:
```diff
@@ -50,14 +50,16 @@ module.exports.loader = ( { preserveCssCustomProperties, includePaths, prelude }
+module.exports.plugins = ( { cssChunkFilename, cssFilename, minify } ) => [
 	new MiniCssExtractPluginWithRTL( {
+		chunkFilename: cssChunkFilename,
 		filename: cssFilename,
```
```suggestion
		filename,
```

sirreal commented on packages/calypso-build/webpack/sass.js:
```diff
@@ -50,14 +50,16 @@ module.exports.loader = ( { preserveCssCustomProperties, includePaths, prelude }
-module.exports.plugins = ( { cssFilename, minify } ) => [
+module.exports.plugins = ( { cssChunkFilename, cssFilename, minify } ) => [
 	new MiniCssExtractPluginWithRTL( {
+		chunkFilename: cssChunkFilename,
```
```suggestion
		chunkFilename,
```

sirreal commented on packages/calypso-build/webpack/sass.js:
```diff
@@ -50,14 +50,16 @@ module.exports.loader = ( { preserveCssCustomProperties, includePaths, prelude }
+ * @return {Object[]}                     styling relevant webpack plugin objects
  */
-module.exports.plugins = ( { cssFilename, minify } ) => [
+module.exports.plugins = ( { cssChunkFilename, cssFilename, minify } ) => [
```
```suggestion
module.exports.plugins = ( { chunkFilename, filename, minify } ) => [
```

sirreal commented on packages/calypso-build/webpack.config.js:
```diff
@@ -53,7 +53,10 @@ function getWebpackConfig(
-	const cssFilename = '[name].css';
+	const cssFilename = path.parse( outputFilename ).name + '.css';
+	const cssChunkFilename = outputChunkFilename
+		? path.parse( outputChunkFilename ).name + '.css'
```
These are "template" names. I don't think we want to use path parsing here. I believe they'll tend to look more like URLs than file system paths.

I'd propose an implementation like what [`SourceMapDevToolPlugin` does](https://github.com/webpack/webpack/blob/de81dd096c4f9d2adb03c9a89207a8a7ad499a6f/lib/SourceMapDevToolPlugin.js#L252-L257).

- Pull off a `?query` if found..
- Change the extension to `css`.
- Append the `?query`.

Changing the extension can likely [expect to find `/\.js$/i`](https://webpack.js.org/configuration/output#outputfilename).

sirreal replied:
Done in f62face1da3872babee1027e4d4c77f09b7c5e3b

ockham replied:
Ah, good point, didn't think of query strings here. (`path.parse().name` was just so temptingly elegant :grimacing: )

sirreal committed:
Implement suggested css filename and chunkFilename

Based on:
https://github.com/webpack/webpack/blob/de81dd096c4f9d2adb03c9a89207a8a7ad499a6f/lib/SourceMapDevToolPlugin.js#L252-L257

See:

https://github.com/Automattic/wp-calypso/pull/32194#discussion_r274285300

sirreal committed:
Remove bad SassConfig includePaths

A bad includePaths had been erroneously brought over from Calypso.
It was accessing the package/client (nonexistent).

simison commented:
I've just started to familiarize with the problem so I might be missing something here;

Does this require us to have JS chunk counterpart, do we _always_ have it? 

Does it make sense to have multiple different CSS chunks async loaded and if so, does this support it?

I know the inspiration came from needing to cache-bust these:

```js
import( /* webpackChunkName: "swiper" */ 'swiper/dist/js/swiper.js' ),
import( /* webpackChunkName: "swiper" */ 'swiper/dist/css/swiper.css' ),
```
[via](https://github.com/Automattic/jetpack/blob/354a493bcb88fd7db71eb318eef1589f890d09e7/extensions/blocks/slideshow/create-swiper.js#L44-L45)

What if we'd have no swiper JS chunk and two CSS imports:
```js
import( /* webpackChunkName: "swiper" */ 'swiper/dist/css/swiper.css' ),

// Further down inside some condition:
import( /* webpackChunkName: "swiper" */ 'swiper/dist/css/swiper-more.css' ),
```

sirreal commented:
> I've just started to familiarize with the problem so I might be missing something here;
> Does this require us to have JS chunk counterpart, do we _always_ have it?
No, it should not.
> Does it make sense to have multiple different CSS chunks async loaded and if so, does this support it?
Yes, and yes it should.

> What if we'd have no swiper JS chunk and two CSS imports:
> 
> ```js
> import( /* webpackChunkName: "swiper" */ 'swiper/dist/css/swiper.css' ),
> 
> // Further down inside some condition:
> import( /* webpackChunkName: "swiper" */ 'swiper/dist/css/swiper-more.css' ),
> ```

We're really working with templates and letting webpack figure out the specific names.

For the no-JS multiple-CSS case, it depends on the provided template, but I'd expect something like the following:
`swiper.zxy987.css`
`swiper-more.abcdef.css`

sirreal committed:
Restructure and test util

sirreal committed:
Rename SASS-specific config options

sirreal commented on packages/calypso-build/webpack/test/util.js:
```diff
@@ -0,0 +1,23 @@
+	} );
+
+	test( 'Returns other extensions untouched 😬', () => {
+		expect( 'static.jsx' ).toBe( 'static.jsx' );
```
I _think_ this is OK. I don't think you ever want to get an extension other than `js` out of webpack.

simison reviewed (approved):
Works beautifully also when I copy Slideshow block to o2 blocks:

<img width="360" alt="image" src="https://user-images.githubusercontent.com/87168/55942813-e3776280-5c4d-11e9-9991-e50537cf6520.png">

I provided extra option in `package.json`'s `build` to get the hash in the name.

sirreal committed:
Update sassConfig option

sirreal commented on packages/calypso-build/webpack/util.js:
```diff
@@ -0,0 +1,17 @@
+function cssNameFromFilename( name ) {
+	if ( name ) {
+		const [ cssChunkFilename, chunkQueryString ] = name.split( '?', 2 );
+		return cssChunkFilename.replace(
```
If there's a query string but the extension isn't `.js`, we return the input without the query:

```
x.ts?query
x.ts
```

That's buggy. Maybe we should throw and error if it doesn't match `/\.js$/i`
